### EDA Section 

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

df = pd.read_csv('../data/raw/train.csv')

In [2]:
#missing values

df.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code      11
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
dtype: int64

In [3]:
#Describe the data set using describe function

df.describe()

,Row ID,Postal Code,Sales
count,9800.000000,9789.000000,9800.000000
mean,4900.500000,55273.322403,230.769059
std,2829.160653,32041.223413,626.651875
min,1.000000,1040.000000,0.444000
25%,2450.750000,23223.000000,17.248000
50%,4900.500000,58103.000000,54.490000
75%,7350.250000,90008.000000,210.605000
max,9800.000000,99301.000000,22638.480000


## Sales Statistics Analysis

1. The dataset contains 9,800 sales transactions.
2. Average sales per transaction is $230.77.
3. Median sales value is $54.49, indicating that half of all transactions are below this amount.
4. The maximum sales value is $22,638.48, significantly higher than the average.
5. The sales distribution appears positively skewed because the mean is much greater than the median.
6. Most transactions are relatively small, while a few high-value transactions contribute substantially to overall revenue.
7. The large gap between the average and maximum sales values suggests the presence of high-value orders or outliers.

In [4]:
## Unique values in each column

df.nunique()

Row ID           9800
Order ID         4922
Order Date       1230
Ship Date        1326
Ship Mode           4
Customer ID       793
Customer Name     793
Segment             3
Country             1
City              529
State              49
Postal Code       626
Region              4
Product ID       1861
Category            3
Sub-Category       17
Product Name     1849
Sales            5757
dtype: int64

## Unique Values Analysis

1. The dataset contains 4,922 unique orders and 9,800 transactions, indicating that a single order can contain multiple products.
2. There are 793 unique customers in the dataset.
3. Sales transactions occurred across 529 cities and 49 states.
4. The business operates in 4 regions.
5. The dataset contains approximately 1,850 unique products.
6. Products are grouped into 3 categories and 17 sub-categories.
7. Customers belong to 3 segments.
8. Four shipping modes are available.
9. The dataset is limited to a single country (United States).
10. Sales values show high variability with 5,757 unique transaction amounts.

In [5]:
### COUNT BY REGION CATEGORY AND SEGMENT

df['Region'].value_counts()

Region
West       3140
East       2785
Central    2277
South      1598
Name: count, dtype: int64

## Region Analysis

1. The West region has the highest number of transactions (3,140).
2. The East region follows with 2,785 transactions.
3. The South region has the lowest number of transactions (1,598).
4. Sales activity appears to be concentrated in the West and East regions.
5. Further analysis is required to determine whether higher transaction volume also results in higher revenue.

## Business Insight
West Region = Largest Market
South Region = Smallest Market

In [6]:
df['Category'].value_counts()

Category
Office Supplies    5909
Furniture          2078
Technology         1813
Name: count, dtype: int64

## Category Analysis

1. Office Supplies is the most frequently purchased category with 5,909 transactions.
2. Furniture accounts for 2,078 transactions.
3. Technology has the lowest transaction count at 1,813.
4. More than half of all transactions belong to the Office Supplies category.
5. High transaction volume does not necessarily imply high revenue or profit, which should be analyzed further.

## Business Insights
Office Supplies drives transaction volume.
Technology may drive revenue despite fewer transactions.

In [7]:
df['Segment'].value_counts()

Segment
Consumer       5101
Corporate      2953
Home Office    1746
Name: count, dtype: int64

## Customer Segment Analysis

1. Consumer is the largest customer segment with 5,101 transactions.
2. Corporate customers account for 2,953 transactions.
3. Home Office is the smallest segment with 1,746 transactions.
4. Consumer customers contribute more than half of the total transaction volume.
5. Further analysis should determine which segment contributes the highest revenue.

## Business Insights

1. Consumer Segment = Largest customer base.
2. Corporate Segment = Potential high-value customers.
3. Home Office = Smallest segment.

### Next EDA Section: Revenue Analysis

In [8]:
##Groupby Region

df.groupby('Region')['Sales'].sum().sort_values(ascending=False)

Region
West       710219.6845
East       669518.7260
Central    492646.9132
South      389151.4590
Name: Sales, dtype: float64

## Regional Sales Analysis

1. The West region generated the highest revenue of $710,219.68.
2. The East region generated the second-highest revenue of $669,518.73.
3. The South region generated the lowest revenue of $389,151.46.
4. The West and East regions together contribute a significant portion of total sales.
5. Revenue generation is unevenly distributed across regions.

In [9]:
df.groupby('Category')['Sales'].sum().sort_values(ascending=False)

Category
Technology         827455.8730
Furniture          728658.5757
Office Supplies    705422.3340
Name: Sales, dtype: float64

## Category Sales Analysis

1. Technology generated the highest revenue of $827,455.87.
2. Furniture generated revenue of $728,658.58.
3. Office Supplies generated revenue of $705,422.33.
4. Although Office Supplies has the highest transaction volume, it generates the lowest revenue among the three categories.
5. Technology contributes the highest revenue despite having the fewest transactions.
6. This indicates that Technology products have a significantly higher average selling price compared to other categories.

In [10]:
df.groupby('Segment')['Sales'].sum().sort_values(ascending=False)

Segment
Consumer       1.148061e+06
Corporate      6.884941e+05
Home Office    4.249822e+05
Name: Sales, dtype: float64

## Customer Segment Sales Analysis

1. The Consumer segment generated the highest revenue of $1,148,061.
2. Corporate customers generated revenue of $688,494.
3. Home Office customers generated revenue of $424,982.
4. Consumer customers represent the largest customer segment in terms of both transaction volume and revenue contribution.
5. Corporate customers contribute significant revenue despite having fewer transactions than Consumer customers.
6. Home Office is the smallest segment in both transaction count and revenue generation.

## Date Analysis

In [11]:
df['Order Date'] = pd.to_datetime(df['Order Date'],dayfirst = True)
df['Ship Date'] = pd.to_datetime(df['Ship Date'], dayfirst = True)

In [12]:
df['Order Date'].min()

Timestamp('2015-01-03 00:00:00')

In [13]:
df['Order Date'].max()

Timestamp('2018-12-30 00:00:00')

In [14]:
df['Ship Date'].min()

Timestamp('2015-01-07 00:00:00')

In [15]:
df['Ship Date'].max()

Timestamp('2019-01-05 00:00:00')

## Date Analysis Observations

1. The dataset covers a period from 03-Jan-2015 to 30-Dec-2018.
2. Sales transactions span approximately four years.
3. The earliest shipment occurred on 07-Jan-2015.
4. The latest shipment occurred on 05-Jan-2019.
5. Ship dates extend beyond the order date range because orders placed at the end of 2018 were delivered in early 2019.
6. The dataset is suitable for time-series analysis, seasonal trend analysis, and year-over-year performance comparisons.

In [16]:
## Orders per year

df['Order Date'].dt.year.value_counts()

Order Date
2018    3258
2017    2534
2016    2055
2015    1953
Name: count, dtype: int64

In [17]:
df['Order Date'].dt.month.value_counts()

Order Date
11    1449
12    1382
9     1354
10     809
5      725
7      697
8      693
6      691
3      680
4      657
1      366
2      297
Name: count, dtype: int64

In [18]:
df.groupby(df['Order Date'].dt.year)['Sales'].sum()

Order Date
2015    479856.2081
2016    459436.0054
2017    600192.5500
2018    722052.0192
Name: Sales, dtype: float64

In [19]:
df['Shipping Days'] = (df['Ship Date']-df['Order Date']).dt.days

In [20]:
df['Shipping Days'].describe()

count    9800.000000
mean        3.961122
std         1.749614
min         0.000000
25%         3.000000
50%         4.000000
75%         5.000000
max         7.000000
Name: Shipping Days, dtype: float64

## Shipping Duration Analysis

1. The average shipping duration is approximately 4 days.
2. The median shipping duration is 4 days, indicating a balanced distribution.
3. The shortest shipping duration is 0 days, suggesting some orders were shipped on the same day.
4. The longest shipping duration is 7 days.
5. 75% of all orders are shipped within 5 days.
6. Shipping performance appears consistent with no significant delays or extreme outliers.
7. The company maintains an efficient fulfillment process with most orders delivered within one week.

In [21]:
df['Shipping Days'].value_counts().sort_index()

Shipping Days
0     514
1     363
2    1295
3     978
4    2718
5    2147
6    1170
7     615
Name: count, dtype: int64

## Shipping Duration Frequency Analysis

1. The most common shipping duration is 4 days, accounting for 2,718 orders.
2. The second most common shipping duration is 5 days with 2,147 orders.
3. A total of 514 orders were shipped on the same day they were placed.
4. Approximately 22% of all orders were shipped within 2 days.
5. Only 615 orders required the maximum shipping duration of 7 days.
6. Most orders are fulfilled within 4 to 5 days, indicating a consistent and efficient shipping process.
7. The shipping distribution suggests stable operational performance with limited delays.

In [22]:
df.groupby(df['Order Date'].dt.month)['Sales'].sum()

Order Date
1      94291.6296
2      59371.1154
3     197573.5872
4     136283.0006
5     154086.7237
6     145837.5233
7     145535.6890
8     157315.9270
9     300103.4117
10    199496.2947
11    350161.7110
12    321480.1695
Name: Sales, dtype: float64

## Monthly Sales Analysis

1. November generated the highest revenue of $350,161.
2. December generated the second-highest revenue of $321,480.
3. September generated the third-highest revenue of $300,103.
4. February recorded the lowest revenue of $59,371.
5. Sales show a clear seasonal pattern, with revenue increasing significantly during Q4.
6. The holiday season (November–December) contributes a substantial portion of total annual revenue.
7. Business demand appears strongest toward the end of the year.

In [23]:
df.groupby('State')['Sales'].sum().sort_values(ascending=False).head(10)

State
California      446306.4635
New York        306361.1470
Texas           168572.5322
Washington      135206.8500
Pennsylvania    116276.6500
Florida          88436.5320
Illinois         79236.5170
Michigan         76136.0740
Ohio             75130.3500
Virginia         70636.7200
Name: Sales, dtype: float64

## State-Level Sales Analysis

1. California generated the highest revenue of $446,306.
2. New York generated the second-highest revenue of $306,361.
3. Texas ranked third with $168,573 in sales.
4. California and New York significantly outperform all other states.
5. Revenue generation is concentrated among a few key states.
6. These states represent important markets for business growth and customer acquisition.
7. Geographic sales performance varies considerably across states.

In [24]:
df.groupby('City')['Sales'].sum().sort_values(ascending=False).head(10)

City
New York City    252462.5470
Los Angeles      173420.1810
Seattle          116106.3220
San Francisco    109041.1200
Philadelphia     108841.7490
Houston           63956.1428
Chicago           47820.1330
San Diego         47521.0290
Jacksonville      44713.1830
Detroit           42446.9440
Name: Sales, dtype: float64

## City-Level Sales Analysis

1. New York City generated the highest revenue of $252,462.
2. Los Angeles ranked second with $173,420 in sales.
3. Seattle, San Francisco, and Philadelphia each generated over $100,000 in revenue.
4. California contributes significantly to overall sales through multiple high-performing cities.
5. Revenue is concentrated among a small number of metropolitan markets.
6. The top five cities generate substantially higher revenue compared to the remaining cities in the top ten.
7. These cities represent key markets for customer acquisition and business growth initiatives.

In [25]:
df.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code      11
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
Shipping Days     0
dtype: int64

In [26]:
df.dtypes

Row ID                    int64
Order ID                    str
Order Date       datetime64[us]
Ship Date        datetime64[us]
Ship Mode                   str
Customer ID                 str
Customer Name               str
Segment                     str
Country                     str
City                        str
State                       str
Postal Code             float64
Region                      str
Product ID                  str
Category                    str
Sub-Category                str
Product Name                str
Sales                   float64
Shipping Days             int64
dtype: object